### DISCUSSION TITLE
A nicer Hyperledger Fabric contract client

### DISCUSSION THREAD

**Discussion by pratikpc** [hyperledger/fabric-sdk-node#567] [General]
In TS we currently would use Fabric Contract API which has a lot of annotations which allows us to write clean, simple and readable code.

However, the code for clients currently in use is barely readable.

```typescript
console.log(
'Using traditional approach to say hi to Hyperledger Fabric',
(
  await contract.evaluateTransaction(
     'sayHi',
     'Pratik'
  )).toString()
);
```

This code has a few issues
1. Needs you to convert output to String
2. Needs you to pass the param as String
3. Needs you to know whether evaluate or submit
4. No intellisense
5. Needs you to remember the name of the chaincode when creating
6. The biggest issue is that all of the annotations and metadata currently remains underutilised by the code.

To fix this, I created my own library 
https://www.npmjs.com/package/@pratikpc/fabric-contract-client

This supports intellisense

![image](https://user-images.githubusercontent.com/28534575/150992668-88621509-4e49-4be2-a07e-a74069d09c26.png)

And allows you to write the following code

```typescript
console.log(
`Chaincode says ${await client.sayHi('Pratik')}`
);
```

This code is much easier to read and handles all conversions on its own.

The best part, both the codebases can be used in the same object.

You can set up the client with 
```typescript
const client = ContractClient(ChainCodeContract, network);
```
The lib extract the name from the Info annotation of the chaincode object or allow you to pass it as Param 3.

I was thinking that this could be useful for fellow Hyperledger Fabric users and wished to merge my library into the fabric-network codebase. Could someone tell me what would the best ways to proceed would be?

**Comment by bestbeforetoday** [comment]
It's a nice approach. Essentially exactly the kind of thing I would be implementing in a real client application, so my general application logic was just interacting with some abstraction layer that internally did appropriate un/marshalling of data and invocation of transaction functions using the SDK.

One suggestion would be to use the metadata that the smart contract API produces rather than using reflection on the smart contract implementation code directly. There's a special transaction function that you can evaluate to obtain the metadata. This would allow you to auto-generate client stubs for smart contracts developed in other languages such as Go or Java, not just JavaScript. See here for some details:

- https://lists.hyperledger.org/g/fabric/topic/85103256
- https://hyperledger.github.io/fabric-chaincode-node/main/api/tutorial-annotated-contract-metadata.html

This seems to me more of a development-/tooling-time package, rather than a runtime package that would naturally fit within the SDK repository itself. Kinda like the gRPC/protobuf compiler, which generates client stubs for specific gRPC services, rather than the core gRPC libraries that deal with the mechanics of runtime service invocation.

That's not to say it couldn't be a Hyperledger / Fabric package in its own right though. It might be worth exploring through the Fabric RFC process: https://github.com/hyperledger/fabric-rfcs

One final thing... looking to future Fabric versions, you might also consider basing generated code on the [@hyperledger/fabric-gateway](https://www.npmjs.com/package/@hyperledger/fabric-gateway) package, which was introduced alongside Fabric v2.4.

**Comment by pratikpc** [comment]
@bestbeforetoday I had a few queries and doubts regarding the comment I wanted to ask you given my lack of experience with the Hyperledger project so I sent a connection request via LinkedIn.

I hope that's alright

**Reply by bestbeforetoday** [reply]
Sorry, I didn't spot which was your LinkedIn request. It would probably be better to have the discussion and ask questions to Hyperledger staff, and Fabric maintainers and contributors in general on Discord:

- Invite link: https://discord.gg/hyperledger
- Fabric contributors channel: https://discord.com/channels/905194001349627914/941396957635682314

You could also share what you've created at a contributor meeting:

- https://wiki.hyperledger.org/display/fabric/Contributor+Meetings

**Comment by bestbeforetoday** [comment]
@pratikpc You can see from the updated [off-chain data store sample](https://github.com/hyperledger/fabric-samples/blob/9f844e5de3195f488797ca87598d7cbefff0c399/off_chain_data/application-typescript/src/contract.ts) I created recently, your generated code is very similar to what I have hand-crafted. If you are still interested in developing and contributing this client code generation tool, I would really encourage you to engage with the Fabric leadership on Discord, or share what you've done and discuss at a contributor meeting (see links [above](https://github.com/hyperledger/fabric-sdk-node/discussions/567#discussioncomment-2924422)).

**Reply by pratikpc** [reply]
Thanks!!
I'll do this

For some reason, I didn't receive a notification for the first reply
